src/ApacheOrcDotNet.OptimizedReader/Infrastructure/HttpByteRangeProvider.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+
+namespace ApacheOrcDotNet.OptimizedReader.Infrastructure
+{
+    public sealed class HttpByteRangeProvider : IByteRangeProvider
+    {
+        readonly HttpClient _httpClient = new();
+        readonly string _remoteLocation;
+
+        internal HttpByteRangeProvider(string remoteLocation)
+        {
+            _remoteLocation = remoteLocation;
+        }
+
+        public void Dispose() => _httpClient.Dispose();
+
+        public int GetRange(Span<byte> buffer, long position)
+        {
+            var request = CreateRangeRequest(position, position + buffer.Length);
+            var response = _httpClient.Send(request);
+
+            if (!response.Content.Headers.ContentRange.Length.HasValue)
+                throw new InvalidOperationException("Range respose must include a length.");
+
+            return DoRead(response.Content.ReadAsStream(), buffer);
+        }
+
+        public async Task<int> GetRangeAsync(Memory<byte> buffer, long position)
+        {
+            var request = CreateRangeRequest(position, position + buffer.Length);
+            var response = await _httpClient.SendAsync(request);
+
+            if (!response.Content.Headers.ContentRange.Length.HasValue)
+                throw new InvalidOperationException("Range respose must include a length.");
+
+            return await DoReadAsync(await response.Content.ReadAsStreamAsync(), buffer);
+        }
+
+        public int GetRangeFromEnd(Span<byte> buffer, long positionFromEnd)
+        {
+            var request = CreateRangeRequest(null, positionFromEnd);
```

Comment:
For calculating that, we would need to know the total length of the data we are getting the ranges from.  Something like:
- _from_: `(totalLength - (bytesFromEnd + buffer.leng))`
- _to_: `<same_value_in_from> + buffer.length`

We were doing a HEAD request to get the total length, but unfortunately, that would be inaccurate for endpoints that do not support HEAD requests. To avoid that and since we only read from the end of the file to get the tail info, the default _x bytes from the end_ request style was kept.

The specs I used for these are [here](https://www.rfc-editor.org/rfc/rfc7233#page-5).

Reply:
I think the length of the provided `Span` should determine how many bytes to read from the end of the file.  You don't need `positionFromEnd`.

Reply:
Yeah. I was considering scenarios like going back 100 bytes and reading the first 50 into the buffer, but we only use it to read entire blocks for the tail. Removed.